### PR TITLE
fix(BUX-644): return stable pub key from PKI capability

### PR DESCRIPTION
**Pull request by arkadiuszos4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/model_xpubs.go [hunk 1]
```diff
@@ -19,8 +19,8 @@ type Xpub struct {
...
 	ID              string `json:"id" toml:"id" yaml:"id" gorm:"<-:create;type:char(64);primaryKey;comment:This is the sha256(xpub) hash" bson:"_id"`
 	CurrentBalance  uint64 `json:"current_balance" toml:"current_balance" yaml:"current_balance" gorm:"<-;comment:The current balance of unspent satoshis" bson:"current_balance"`
-	NextInternalNum uint32 `json:"next_internal_num" toml:"next_internal_num" yaml:"next_internal_num" gorm:"<-;type:int;comment:The next index number for the internal xPub derivation" bson:"next_internal_num"`
-	NextExternalNum uint32 `json:"next_external_num" toml:"next_external_num" yaml:"next_external_num" gorm:"<-;type:int;comment:The next index number for the external xPub derivation" bson:"next_external_num"`
+	NextInternalNum uint32 `json:"next_internal_num" toml:"next_internal_num" yaml:"next_internal_num" gorm:"<-;type:int not null;default:0comment:The next index number for the internal xPub derivation" bson:"next_internal_num"`
+	NextExternalNum uint32 `json:"next_external_num" toml:"next_external_num" yaml:"next_external_num" gorm:"<-;type:int not null;default:1;comment:The next index number for the external xPub derivation" bson:"next_external_num"`
```
Comment (dorzepowski) [inline engine/model_xpubs.go:23]: Maybe I'm wrong, or missing some piece,
but I think that having separated `xpub.next_internal_num`, `xpub.next_external_num`, `paymail.external_xpub_key_nume` and `paymail.pub_key_num` can probably lead to some collisions (reusage of the derived keys)

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: engine/action_xpub.go [hunk 2]
```diff
@@ -19,6 +19,8 @@ func (c *Client) NewXpub(ctx context.Context, xPubKey string, opts ...ModelOps)
 		xPubKey, c.DefaultModelOptions(append(opts, New())...)...,
 	)
 
+	xPub.NextExternalNum = 1 // 0 is reserved for paymail adress external Xpub
```
Comment (chris-4chain) [inline engine/action_xpub.go:22]: For developers less familiar with this - like myself - there should also be a comment 'Why 1.?'"

[the remark this answers is not in the document]
Comment (arkadiuszos4chain) [reply engine/action_xpub.go]: not relevant anymore

File: engine/model_xpubs.go (hunk 1 above)
Comment (arkadiuszos4chain) [reply engine/model_xpubs.go]: Fixed/changed as we discussed it

File: engine/model_paymail_addresses.go [hunk 3]
```diff
@@ -178,7 +187,65 @@ func (m *PaymailAddress) GetExternalXpub() (*bip32.ExtendedKey, error) {
...
+	pubKey, err := hdPubKey.ECPubKey()
+	if err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(pubKey.SerialiseCompressed()), nil
+}
+
+func (m *PaymailAddress) GetNextXpub(ctx context.Context) (*bip32.ExtendedKey, error) {
+	lockKey := fmt.Sprintf(lockKeyProcessXpub, m.XpubID)
```
[the remark this answers is not in the document]
Comment (arkadiuszos4chain) [reply engine/model_paymail_addresses.go]: Moved to locks.go

File: engine/action_xpub.go (hunk 2 above)
Comment (arkadiuszos4chain) [reply engine/action_xpub.go]: Not relevant anymore

File: engine/model_paymail_addresses.go [hunk 4]
```diff
@@ -159,8 +164,12 @@ func (m *PaymailAddress) GetIdentityXpub() (*bip32.ExtendedKey, error) {
...
-// GetExternalXpub will get the external xPub
-func (m *PaymailAddress) GetExternalXpub() (*bip32.ExtendedKey, error) {
+// getExternalXpub will get the external xPub
+func (m *PaymailAddress) getExternalXpub() (*bip32.ExtendedKey, error) {
+	if m.externalHdXpub != nil {
+		return m.externalHdXpub, nil
+	}
+
 	// Check if the xPub was encrypted
 	if len(m.ExternalXpubKey) != utils.XpubKeyLength {
```
[the remark this answers is not in the document]
Comment (arkadiuszos4chain) [reply engine/model_paymail_addresses.go:176]: It's not a typical check, and certainly not a common scenario for a string (xpub) with a length other than `utils.XpubKeyLength` to be decrypted. Therefore, I would prefer to keep it in place.

[the remark this answers is not in the document]
Comment (chris-4chain) [reply engine/model_paymail_addresses.go:176]: If it's not a typical check then you're totally right 👍

**Review by Nazarii-4chain — APPROVED** [review]

File: engine/model_paymail_addresses_test.go [hunk 5]
```diff
@@ -112,4 +118,192 @@ func TestNewPaymail(t *testing.T) {
 		address1, _ := bitcoin.GetAddressFromPubKey(key1, true)
 		assert.Equal(t, addressInternal, address1.AddressString)
 	})
+
+	t.Run("ctor test", func(t *testing.T) {
```
Comment (dorzepowski) [inline engine/model_paymail_addresses_test.go:122]: What does it mean `ctor`?

File: engine/model_destinations.go [hunk 6]
```diff
@@ -20,14 +20,15 @@ type Destination struct {
...
+	LockingScript                string `json:"locking_script" toml:"locking_script" yaml:"locking_script" gorm:"<-:create;type:text;comment:This is Bitcoin output script in hex" bson:"locking_script"`
+	Type                         string `json:"type" toml:"type" yaml:"type" gorm:"<-:create;type:text;comment:Type of output" bson:"type"`
+	Chain                        uint32 `json:"chain" toml:"chain" yaml:"chain" gorm:"<-:create;type:int;comment:This is the (chain)/num location of the address related to the xPub" bson:"chain"`
+	Num                          uint32 `json:"num" toml:"num" yaml:"num" gorm:"<-:create;type:int;comment:This is the chain/(num) location of the address related to the xPub" bson:"num"`
+	PaymailExternalDerivationNum uint32 `json:"paymail_external_derivation_num" toml:"paymail_external_derivation_num" yaml:"paymail_external_derivation_num" gorm:"<-:create;type:int not null;comment:This is the chain/(num)/(ext_derivation_num) location of the address related to the xPub" bson:"paymail_external_derivation_num"`
```
Comment (dorzepowski) [inline engine/model_destinations.go:29]: ⚠️ this change require changes in go-client and js-client ⚠️ 
We shouldn't merge it before changes clients will be prepared also

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: engine/model_destinations.go (hunk 6 above)
Comment (arkadiuszos4chain) [reply engine/model_destinations.go:29]: Why? Do we even need key derivation information at client side?

Comment (arkadiuszos4chain) [reply engine/model_destinations.go]: go-client : https://github.com/bitcoin-sv/spv-wallet-go-client/pull/204
js-client: https://github.com/bitcoin-sv/spv-wallet-js-client/pull/888

**Review by dorzepowski — APPROVED** [review]

**Pull request merged**